File: Fase 2_3/Client/src/Thash.py
import os


class NodeHash:
  def __init__(self, id, name) -> None:
    self.id = id
    self.name = name


class Hash:
  def __init__(self, m:int , min:int, max:int):
    # min,max ,n ,m= 0
    self.n = 0
    self.iterator = 0
    self.h = []
    self.m = m
    self.min = min
    self.max = max
    self.init()

  def division(self, k:int):
    return (int(k%self.m))

  def linear(self, k:int):
    return (int((k%3)+1)*self.iterator)
    # return (int((k+1)%self.m))

  def init(self):
    self.iterator = 0
    self.n = 0
    self.h = [None]*int(self.m)
    for i in range(int(self.m)):
      self.h[i] = NodeHash(-1,"")

  def insert(self, k:int , name:str):
    i = self.division(k)
    if (self.h[i].id == -1):
      if ((self.n*100/self.m) >= self.max):
        self.rehashing()
      else:
        # nodo = new NodeHash(k,name)
        self.h[i] = NodeHash(k,name)
        self.n+=1

    else:
      # self.n+=1
      self.iterator +=1
      nuevo = self.linear(k)
      nuevo = nuevo + self.iterator
      if(nuevo <= len(self.h)):
        if(self.h[nuevo].id == -1):
          # self.n+=1
          if ((self.n*100/self.m) >= self.max):
            self.rehashing()
          self.h[nuevo] = NodeHash(k,name)
          self.n+=1
        else:
          self.rehashing()
          # self.h[nuevo] = k
      else:
        self.rehashing()
        if ((self.n*100/self.m) >= self.max): self.rehashing()
        else:self.h[nuevo] = NodeHash(k,name)

    self.sout()
  # def insert(self, k:int):
  #   i = self.division(k)
  #   i = int(i)
  #   while (self.h[i] != -1):
  #     i = self.linear(i)
  #   self.h[i] = k
  #   self.n +=1
  #   self.rehashing()


  def rehashing(self):
    #Array copy
    # self.sout()
    temp = self.h
    # Rehash now
    mprev = self.m
    self.m = self.n*100/self.min
    self.init()
    for i in range(int(mprev)):
      if (temp[i].id != -1):
        self.insert(temp[i].id, temp[i].name)
    # self.sout()

  def sout(self):
    print('')
    # print('[',end=' ')
    # for i in range(int(self.m)):
    #   print(' '+str(self.h[i].id),end=' ')
    # print(']',self.n*100/self.m,'%')

  def grafica(self):
    text = '''digraph {
  node [ shape=none fontname=Helvetica ]\nn3 [ label = <
    <table border="1">
       <tr><td colspan="1" bgcolor="Peru">id</td><td colspan="1" bgcolor="Peru">name</td><td colspan="1" bgcolor="Peru">index</td></tr>'''
       
    for i in range(int(self.m)):
      if(self.h[i].id != -1):
        text+='<tr><td bgcolor="orange">'+str(self.h[i].id)+'</td><td bgcolor="orange">'+self.h[i].name+' </td><td bgcolor="yellow">'+str(i)+ '</td></tr>\n\t'

    text+='''\t</table> \n>]\n}'''
    dot = "/home/ijosuer/Escritorio/EDD_2S_BatallaNaval_202000895/archivos/{}_dot.txt".format('Tabl')
    with open(dot, 'w') as grafo:
        grafo.write(text)
    result = "/home/ijosuer/Escritorio/EDD_2S_BatallaNaval_202000895/archivos/{}.png".format('TablaHash')
    os.system("dot -Tpng " + dot + " -o " + result)
    # print(text)
if __name__ == "__main__":
  hashito = Hash(5,20,80)
  hashito.insert(5,"josue")
  hashito.insert(10,"dani")
  # hashito.insert(15,"kpngi")
  # hashito.insert(2,"luca")
  # hashito.insert(20,"jej")
  # hashito.insert(25,"vomit")
  # hashito.insert(30,"skrim")
  hashito.grafica()
  # hashito.insert(35)
  # hashito.insert(40)
  # hashito.insert(45)
  # hashito.insert(50)
  # hashito.insert(55)

File: Fase 2_3/Server/src/THash.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class hashi
{
private:
  /* data */
public:

  int n,m,min,max;
  // vector<int>lista;
  int iterar = 0;

  hashi(int _m , int _min , int _max ){
    m = _m;
    min = _min;
    max = _max;
    init();
  }
  int lista[5];
  void init(){
    n = 0;
    // lista[m];
    for (int i = 0; i < m; i++)
    {
      lista[i] =-1;
    }
    iterar = 0;
  }

  void insert(int k){
    int i = division(k);
    int r = 0;
    if(lista[i] == -1){
      lista[i] = k;
    }else{
      iterar +=1;
      r = linear(k);
      lista[r] = k;
    }
    // int i = division(k);
    // int r =0;
    // while (lista[i] != -1)
    // {
    //   iterar +=1;
    //   i =  linear(i);
    // }
    // lista[i] = k;
    n++;
    rehashing();
    // print();
  }

  void rehashing(){
    int len = (n*100)/m;
    if (len >= max)
    {
      // Array copy
      // int temp[sizeof(lista)/lista[0]];
      int temp[m];
      for (int i = 0; i < sizeof(lista)/sizeof(lista[0]); i++)
      {
        temp[i] = lista[i];
      }
      
      print();

      // Rehashing
      int mprev = m;
      m = n*100/min;
      init();
      cout<<sizeof(lista)/sizeof(lista[0])<<endl;
      for (int i = 0; i < mprev; i++)
      {
        if (temp[i] != -1)
        {
          insert(temp[i]);
        }
      }
      cout<<sizeof(lista)/sizeof(lista[0])<<endl;
    }else{
        print();
      }
  }

  void print(){
    cout<<"[";
    for (int i = 0; i < m; i++)
    {
      cout<<" "<<lista[i];
    }
      cout<<" ] "<<n*100/m<<"%"<<endl;
  }

  int division(int k){
    return k%m;
  }

  int linear(int k){
    return ((k%3)+1)*iterar;
    // return ((k+1)%m);
  }
};


// int main(){
//   hashi hash(5,20,80);
  // int lista[11];
  // cout<<sizeof(lista)/sizeof(lista[0])<<endl;
  // try
  // {
  // hash.insert(0);
  // hash.insert(1);
  // hash.insert(2);
  // hash.insert(3);
  // hash.insert(5);
  // hash.insert(4);
  // hash.insert(5);
  // hash.insert(6);
  // hash.insert(7);
  // hash.insert(8);
  // hash.insert(9);
  // hash.insert(10);
  // hash.insert(11);
  // hash.insert(12);
  // hash.insert(13);
  // hash.insert(26);
  // hash.insert(39);
  // hash.insert(52);
  // hash.insert(65);
  // hash.insert(23);
  // hash.insert(9);
  // hash.insert(8);
  // hash.insert(50);

  // hash.insert(5);
  // hash.insert(10);
  // hash.insert(15);
  // hash.insert(20);
  // hash.insert(25);
  // hash.insert(30);
//   }
//   catch(const std::exception& e)
//   {
//     std::cerr << e.what() << '\n';
//   }
//   return 1;
// }
